astropy/cosmology/core.py

```diff
@@ -166,23 +335,96 @@ def __eq__(self, other):
+        and :math:`\Omega_k=0` and therefore flat, like ``FlatLambdaCDM``.
+
+        Parameters
+        ----------
+        other : `~astropy.cosmology.Cosmology` subclass instance
+            The object in which to compare.
 
+        Returns
+        -------
+        bool
+            True if cosmologies are equivalent, False otherwise.
         """
-        if not isinstance(other, Cosmology):
-            return False
+        # The options are: 1) same class & parameters; 2) same class, different
+        # parameters; 3) different classes, equivalent parameters; 4) different
+        # classes, different parameters. (1) & (3) => True, (2) & (4) => False.
 
-        sias, oias = self._init_arguments, other._init_arguments
+        # [Some check here!] this missing logic is why this method is abstract.
+        # for an example, see `FLRW.equivalent` or `FlatFLRWMixin.equivalent`.
+        # Below is strict equality of the class and parameters.
 
-        # check if the cosmologies have identical signatures.
-        # this protects against one cosmology having a superset of input
-        # parameters to another cosmology.
-        if (sias.keys() ^ oias.keys()) - {'meta'}:
+        # check (4)
+        if other.__class__ is not self.__class__:
             return False
 
-        # are all the non-excluded immutable arguments equal?
-        return all((np.all(oias[k] == v) for k, v in sias.items()
-                    if k != "meta"))
+        # check (1) & (2)
+        sps, ops = self.parameters, other.parameters
+        if (sps.keys() ^ ops.keys()):  # should never be tripped
+            params_eq = False
+        else:
+            params_eq = all((np.all(ops[k] == v) for k, v in sps.items()))
+
+        return params_eq
+
+    # -----------------------------------------------------
+
+    def __repr__(self):
+        # trick for getting dict of Parameter objects instead of the values
+        cps = self.__class__.parameters.fget(self.__class__)
+        ps = self.parameters
+
+        # make formatter strings (see Parameter.__format__) & fill from values
+        namelead = f"{self.__class__.__qualname__}("
+        if self.name is not None:
+            namelead += f"name=\"{self.name}\", "
+        fmtps = (format(v, "None" if ps[k] is None else "") for k, v in cps.items())
+        # ↑ handles if any value is None
+        retstr = namelead + ", ".join(fmtps) + ")"
+        return retstr.format(**ps)
+
+
+class FlatCosmologyMixin(metaclass=abc.ABCMeta):
+    """
+    Base class for Flat mixins. Do not instantiate.
+    Note that all instances of ``FlatCosmologyMixin`` are flat, but not all
+    flat cosmologies are instances of ``FlatCosmologyMixin``. As example,
+    ``LambdaCDM`` **may** be flat (for the a specific set of parameter values),
+    but ``FlatLambdaCDM`` **will** be flat.
+    """
+
+    @abc.abstractmethod
+    def equivalent(self, other):
+        r"""Check equivalence between Cosmologies.
```

Comment:
Why duplicate the docstring? Also, if the point is not to allow this to be called, best to `raise NotImplementedError`?

Reply:
I've duplicated the docstring because the ``Mixin`` go before the ``Cosmology`` in the class inheritance. But I can just replace it with ``equivalent.__doc__ = Cosmology.equivalent.__doc__``, which is much cleaner!

Reply:
Also, implemented ``raise NotImplementedError``.